### ISSUE TITLE
ApplicationModuleTest in Nested JUnit Test fails

### ISSUE THREAD

**Issue by chhorz** [spring-projects/spring-modulith#725]
Hello,

i recognized a weired behaviour when executing JUnit tests using `@ApplicationModuleTest`. The following example is a simplified demo to reproduce the error. In this example i am using version 1.2.1 of the modulith project.

If i execute the tests the innner test within the nested class fails. If i uncomment the test case in the outer class both tests are successful. The demo is a simple project with just one module (link to the demo project at the end of the issue).

```java
@ApplicationModuleTest
@Import(TestcontainersConfiguration.class)
public class AbstractCustomerTest {
}


@Sql(executionPhase = BEFORE_TEST_METHOD, scripts = {"classpath:sql/customer.sql"})
class CustomerServiceTest extends AbstractCustomerTest {

	@Autowired
	private CustomerService service;

//	@Test
//	void getCustomer() {
//		assertThat(service.get(1L))
//				.isNotNull()
//				.satisfies(c -> {
//					assertThat(c.id()).isNotNull();
//					assertThat(c.name()).isEqualTo("John Doe");
//				});
//	}


	@Nested
	class Slice {

		@Test
		void getCustomer() {
			assertThat(service.get(1L))
					.isNotNull()
					.satisfies(c -> {
						assertThat(c.id()).isNotNull();
						assertThat(c.name()).isEqualTo("John Doe");
					});
		}

	}

}

```

A complete sample application can be found here: https://github.com/chhorz/modulith-test-demo

**Comment by odrotbohm** [comment]
My initial investigation shows that for `@Nested` classes, a different kind of `ApplicationContext` arrangement is in play. One is created for the main class. That will get the schema creation applied. For the `@Nested` class, a new `ApplicationContext` instance gets created, the schema gets applied to the new `DataSource`. The initializing script, however, is then executed against the `DataSource` instance of the *first* `ApplicationContext`. That makes sense to some degree, as the annotation lives on the class declaration the `AC` is created for.

In other words, I don't think this is a problem of Spring Modulith, but the question of how Spring's SQL initialization functionality interacts with nested `ApplicationContext` instances. On the other hand, I wonder if that second `AC` should be created at all, i.e., whether we need to fix something about the cache key created for `@ApplicationModuleTest`.

@sbrannen – Would you mind taking a look? Under what circumstances would you expect a new `AC` to be created for `@Nested` classes? If the additional `AC` creation *is* expected, how would one declare a SQL script to be executed against the `DataSource`s created in these nested `AC`s?

**Comment by sbrannen** [comment]
I have not yet had a chance to look at / debug the running application, but I can provide a few pointers in the interim.

> In other words, I don't think this is a problem of Spring Modulith, but the question of how Spring's SQL initialization functionality interacts with nested `ApplicationContext` instances. 

The Javadoc for `@Sql` states: _This annotation will be inherited from an enclosing test class by default. See `@NestedTestConfiguration` for details._

> On the other hand, I wonder if that second `AC` should be created at all, i.e., whether we need to fix something about the cache key created for `@ApplicationModuleTest`.

My hunch is that the second `ApplicationContext` should not be created.

> @sbrannen – Under what circumstances would you expect a new `AC` to be created for `@Nested` classes?

Only if the `MergedContextConfiguration` is different for the enclosing class and nested class.

Thus, I'd turn on DEBUG or TRACE logging for the TestContext framework to see what the cache keys are (i.e., the `MergedContextConfiguration` values).

It may be that Spring Modulith needs to makes use of `TestContextAnnotationUtils` to resolve the issue, but that's just a guess (without having debugged the code).

Let me know what you find out about the context cache keys and whether you still need me to dig deeper.

Cheers,

Sam

**Comment by odrotbohm** [comment]
Thanks! That's been very helpful, Sam! I've diagnosed a broken `equals(…)`/`hashCode()` in some of the infrastructure feeding into the context customizers. Fix coming.

**Comment by odrotbohm** [comment]
This is fixed for 1.3 now. Will be back-ported to 1.2 and 1.1 as well.

**Comment by sbrannen** [comment]
> Thanks! That's been very helpful, Sam! 

🙇 

> I've diagnosed a broken `equals(…)`/`hashCode()` in some of the infrastructure feeding into the context customizers.

Yep, those are often the cause for context cache hit/miss issues.

Glad you sorted it out!
